runtime/src/accounts_cache.rs

```diff
@@ -147,14 +147,27 @@ impl AccountsCache {
     }
 
     pub fn add_root(&self, root: Slot) {
-        self.maybe_unflushed_roots.write().unwrap().insert(root);
+        let max_flushed_root = self.fetch_max_flush_root();
+        if root > max_flushed_root || (max_flushed_root == 0 && root == 0) {
```

Comment:
nits: `root > max_flushed_root || (root == max_flushed_root && root == 0)` (hint: keep the structure for the general case as much as possible)